### PR TITLE
[Pipe] Adds support for head requests

### PR DESCRIPTION
**Pull request by eliasjpr**
Issue https://github.com/amberframework/amber/issues/233

Currently head requests are returning 404.

- Does not render response.body if the request is of type HEAD
- Converts route.cr from class to struct this should give a performance
improvement
- Changes :get symbol to "get"

With the changes it will be possible to make HEAD requests and the
response.body should be empty

Example:

 ○ http -v HEAD :3000

Request
----------------------------
HEAD / HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: localhost:3000
User-Agent: HTTPie/0.9.9

Response
----------------------------
HTTP/1.1 200 OK
Connection: Keep-Alive
Content-Length: 0
Keep-Alive: timeout=5, max=10000
Set-Cookie: amber.session=eyJfZmxhc2giOiJ7fSJ9--mpOMtKPrJi1R9nRQyYtmlwkX5CY%3D; path=/; HttpOnly
X-Powered-By: Amber

✋ @docelic

### CODE REVIEW

File: spec/amber/router/pipe/pipeline_spec.cr [hunk 1]
```diff
@@ -67,6 +67,13 @@ module Amber
           response.body.should eq "Destroy"
         end
 
+        it "perform HEAD request" do
```
Comment (marksiemers) [inline spec/amber/router/pipe/pipeline_spec.cr]: I think this should be `it "responds to HEAD requests"` or `it "provides response for HEAD requests"`

"Performing the request" is more what the client is doing.

File: spec/amber/router/pipe/pipeline_spec.cr [hunk 2]
```diff
@@ -67,6 +67,13 @@ module Amber
...
+          request = HTTP::Request.new("HEAD", "/hello/1")
+          response = create_request_and_return_io(pipeline, request)
+          response.headers["Content-Length"].should eq "0"
+          response.body.empty?.should be_true
+        end
```
Comment (marksiemers) [inline spec/amber/router/pipe/pipeline_spec.cr]: Should we check a positive (200) response when expected and a 404 code as well?

This may end up being scope creep for this PR, as these comments could apply to many of these tests.

File: src/amber/router/route.cr [hunk 3]
```diff
@@ -1,5 +1,5 @@
 module Amber
-  class Route
+  struct Route
```
Comment (marksiemers) [inline src/amber/router/route.cr]: This seems like a big change, and I'm not sure it directly relates to the HEAD requests, does it?

Comment (eliasjpr) [reply src/amber/router/route.cr]: This was an improvement that I've been wanting to make. This has to do more with the inner workings of crystal. In short, the Route is just holding a struct of data.

For a better understanding of this change read this https://crystal-lang.org/docs/guides/performance.html

File: spec/amber/router/pipe/pipeline_spec.cr (hunk 2 above)
Comment (eliasjpr) [reply spec/amber/router/pipe/pipeline_spec.cr]: In that case we should check for all headers, this was more checking if the pipeline was processing the content.

File: src/amber/router/route.cr (hunk 3 above)
Comment (marksiemers) [reply src/amber/router/route.cr]: I've read about the performance gains before, a struct goes on the stack and is immutable (or copied by value anyway).

So I think it will be an improvement, but I ran into issues when working on a caching solution where I tried a struct, but it didn't work because I needed parts to be mutable.

I suppose a route should never change during run time - it just seems like a non-zero risk, and putting it in here means if we rollback, then we rollback the HEAD functionality too.

File: spec/amber/router/pipe/pipeline_spec.cr (hunk 2 above)
Comment (marksiemers) [reply spec/amber/router/pipe/pipeline_spec.cr]: I see, I was muddying the waters between pipeline tests and controller tests. I should have looked closer at the naming.

**Comment by docelic** [comment]
That way would prevent headers from being added, if a GET would set some.

**Comment by eliasjpr** [comment]
@robacarp the head request cannot be hijack since the user my be setting headers during the process of an action. The metainformation contained in the HTTP headers in response to a HEAD request SHOULD be identical to the information sent in response to a GET request For the given path.

**Comment by eliasjpr** [comment]
@marksiemers removed the `struct` from routes.cr opening a PR just for that.
